bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/discovery/RokuDiscoveryService.java

```diff
@@ -0,0 +1,267 @@
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(true);
+        }
+        this.scheduledFuture = null;
+    }
+
+    @Override
+    public void startScan() {
+        doNetworkScan();
+    }
+
+    /**
+     * Enumerate all network interfaces, send the discovery broadcast and process responses.
+     *
+     */
+    private synchronized void doNetworkScan() {
+        try {
+            Enumeration<NetworkInterface> nets = NetworkInterface.getNetworkInterfaces();
+            while (nets.hasMoreElements()) {
+                NetworkInterface ni = nets.nextElement();
+                DatagramSocket socket = sendDiscoveryBroacast(ni);
+                if (socket != null) {
+                    scanResposesForKeywords(socket);
+                }
+            }
+        } catch (IOException e) {
+            logger.debug("Error discovering devices", e);
+        }
+    }
+
+    /**
+     * Broadcasts a SSDP discovery message into the network to find provided services.
+     *
+     * @return The Socket where answers to the discovery broadcast arrive
+     */
+    private @Nullable DatagramSocket sendDiscoveryBroacast(NetworkInterface ni) {
+        try {
+            InetAddress m = InetAddress.getByName("239.255.255.250");
+            final int port = 1900;
+
+            if (!ni.isUp() || !ni.supportsMulticast()) {
+                return null;
+            }
+
+            Enumeration<InetAddress> addrs = ni.getInetAddresses();
+            InetAddress a = null;
+            while (addrs.hasMoreElements()) {
+                a = addrs.nextElement();
+                if (a instanceof Inet4Address) {
+                    break;
+                } else {
+                    a = null;
+                }
+            }
+            if (a == null) {
+                logger.debug("No ipv4 address on {}", ni.getName());
+                return null;
+            }
+
+            // Create the discovery message packet
+            byte[] requestMessage = ROKU_DISCOVERY_MESSAGE.getBytes(StandardCharsets.UTF_8);
+            DatagramPacket datagramPacket = new DatagramPacket(requestMessage, requestMessage.length, m, port);
+
+            // Create socket and send the discovery message
+            DatagramSocket socket = new DatagramSocket();
+            socket.setSoTimeout(3000);
+            socket.send(datagramPacket);
+            return socket;
+        } catch (IOException e) {
+            logger.debug("sendDiscoveryBroacast() got IOException: {}", e.getMessage());
+            return null;
+        }
+    }
+
+    /**
+     * Scans all messages that arrive on the socket and process those that come from a Roku.
+     *
+     * @param socket The socket where answers to the discovery broadcast arrive
+     */
+    private void scanResposesForKeywords(DatagramSocket socket) {
+        boolean socketTimeout = false;
+        do {
+            byte[] receiveData = new byte[1024];
```

Comment:
Do you need a new byte block here? or can you reuse it by putting it above the do-while loop?